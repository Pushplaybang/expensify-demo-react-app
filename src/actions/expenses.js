import uuid from "uuid";
import moment from "moment";
import { db } from "./../firebase/firebase";
/**
 * Synchronous Actions
 * ---
 * 1. component calls action generator
 * 2. action generator returns an object
 * 3. component calls dispatch with the object on the store
 * 4. redux store changes
 *
 * Asynchronous Actions
 * ---
 * 1. component calls action generator
 * 2. action generator returns a function
 * 3. component calls dispatch with the function on the store (not default behavior
 *  requires redux-thunk middleware)
 * 4. function runs, has the ability to dispatch other actions or whatever it wants
 *     - async action - http request etc...
 *     - ends by interacting with redux store via standard action.
 */

// ADD_EXPENSE
export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return dispatch => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return db
      .ref("expenses")
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

// REMOVE_EXPENSE
export const removeExpense = ({ id }) => ({
  type: "REMOVE_EXPENSE",
  id
});

export const startRemoveExpense = ({ id }) => {
  return dispatch => {
    return db.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }));
    });
  };
};

export const setExpenses = expenses => ({
  type: "SET_EXPENSES",
  expenses
});

export const startSetExpenses = () => {
  return dispatch => {
    return db
      .ref("expenses")
      .once("value")
      .then(snapshot => {
        const expenses = [];
        snapshot.forEach(item => {
          expenses.push({
            id: item.key,
            ...item.val()
          });
        });

        dispatch(setExpenses(expenses));
      });
  };
};
