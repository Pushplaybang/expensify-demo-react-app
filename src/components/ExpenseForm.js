import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ""
    };
  }

  onDescriptionChange = event => {
    const description = event.target.value;
    this.setState(() => ({ description }));
  };

  onAmountChange = event => {
    const amount = event.target.value;
    if (amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = createdAt => {
    createdAt && this.setState(() => ({ createdAt }));
  };

  onFocusCalendar = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onNoteChange = event => {
    const note = event.target.value;
    this.setState(() => ({ note }));
  };

  onSubmit = event => {
    event.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: "Please provide an amount and description"
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        note: this.state.note,
        createdAt: this.state.createdAt.valueOf()
      });
    }
  };
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            value={this.state.description}
            onChange={this.onDescriptionChange}
            placeholder="description"
            name="description"
            autoFocus
          />
          <input
            type="text"
            placeholder="amount"
            name="amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusCalendar}
            numberOfMonths={1}
            isOutsideRange={day => false}
          />
          <textarea
            placeholder="add a note for your expense (optional)"
            name="note"
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <button type="submit">
            {this.props.expense ? "Update" : "Add"} Expense
          </button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
