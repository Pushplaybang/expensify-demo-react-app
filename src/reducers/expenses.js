export const expensesReducer = (state = [], action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];
    case "EDIT_EXPENSE":
      return state.map(expense => {
        if (expense.id !== action.id) return expense;
        return { ...expense, ...action.updates };
      });
    case "REMOVE_EXPENSE":
      return state.filter(expense => expense.id !== action.id);
    case "SET_EXPENSES":
      return action.expenses;
    default:
      return state;
  }
};
