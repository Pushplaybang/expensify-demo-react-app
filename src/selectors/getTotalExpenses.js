// sum an array
const reducer = (accumulator, currentValue) => accumulator + currentValue;

export default (expenses = []) => {
  return expenses.map(expense => expense.amount).reduce(reducer, 0);
};
