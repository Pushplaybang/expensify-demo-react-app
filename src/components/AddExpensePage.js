import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm.js";
import { startAddExpense } from "./../actions/expenses.js";

export class AddExpensePage extends React.Component {
  onSubmit = expense => {
    this.props.addExpense(expense);
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addExpense: expense => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
