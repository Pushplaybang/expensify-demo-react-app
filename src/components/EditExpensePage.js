import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm.js";
import { editExpense, startRemoveExpense } from "./../actions/expenses.js";

export class EditExpensePage extends React.Component {
  onSubmit = expense => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push("/");
  };

  onRemove = () => {
    this.props.removeExpense({ id: this.props.expense.id });
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <h1>Edit Expense</h1>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onRemove}>remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense:({id}) => dispatch(startRemoveExpense({ id }))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
