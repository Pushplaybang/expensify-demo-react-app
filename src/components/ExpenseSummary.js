import React from "react";
import { connect } from "react-redux";
import numeral from "numeral";

import getVisibleExpenses from "./../selectors/getVisibleExpenses";
import getTotalExpenses from "./../selectors/getTotalExpenses";

export const ExpenseSummary = ({ count, amount }) => {
  const expenseWord = count === 1 ? 'expense' : 'expenses';
  return (
    <h3>
      Viewing {count || 0} {expenseWord} totalling 
      <strong> {numeral(amount / 100).format("$0,0.00")}</strong>
    </h3>
  );
};

const mapStateToProps = (state, props) => ({
  count: getVisibleExpenses(state.expenses, state.filters).length,
  amount: getTotalExpenses(getVisibleExpenses(state.expenses, state.filters))
});

export default connect(mapStateToProps)(ExpenseSummary);
