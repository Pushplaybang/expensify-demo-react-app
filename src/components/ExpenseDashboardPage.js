import React from "react";

import ExpenseList from "./ExpenseList.js";
import ExpenseListFilters from "./ExpenseListFilters.js";
import ExpenseSummary from "./ExpenseSummary.js";

export const ExpenseDashboardPage = props => (
  <div>
    <ExpenseSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);
