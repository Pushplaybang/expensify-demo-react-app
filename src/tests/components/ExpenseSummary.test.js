import React from "react";
import { shallow } from "enzyme";

import getTotalExpenses from "./../../selectors/getTotalExpenses";
import expenses from "./../fixtures/expenses";

import { ExpenseSummary } from "./../../components/ExpenseSummary";

test("Should render ExpenseSummary with no expenses", () => {
  const wrapper = shallow(<ExpenseSummary />);
  expect(wrapper).toMatchSnapshot();
});

test("Should render ExpenseSummary with one expense", () => {
  const total = getTotalExpenses([expenses[0]]);
  const count = 1;
  const wrapper = shallow(<ExpenseSummary count={count} amount={total} />);

  expect(wrapper).toMatchSnapshot();
});

test("Should render ExpenseSummary with multiple expenses", () => {
  const total = getTotalExpenses(expenses);
  const count = expenses.length;
  const wrapper = shallow(<ExpenseSummary count={count} amount={total} />);

  expect(wrapper).toMatchSnapshot();
});
