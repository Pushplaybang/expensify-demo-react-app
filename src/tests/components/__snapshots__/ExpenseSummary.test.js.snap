// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseSummary with multiple expenses 1`] = `
<h3>
  Viewing 
  3
   
  expenses
   totalling
  <strong>
     
    $6.00
  </strong>
</h3>
`;

exports[`Should render ExpenseSummary with no expenses 1`] = `
<h3>
  Viewing 
  0
   
  expenses
   totalling
  <strong>
     
    $0.00
  </strong>
</h3>
`;

exports[`Should render ExpenseSummary with one expense 1`] = `
<h3>
  Viewing 
  1
   
  expense
   totalling
  <strong>
     
    $1.00
  </strong>
</h3>
`;
